### PR TITLE
fix: [CO-491] caldav schedule agent client and none

### PR DESCRIPTION
**Pull request by frisonisland**
### Goal
When the client sends a CalDAV request with SCHEDULE-AGENT=CLIENT, the server should not send a message for the meeting.

### What's changed

* Added support for SCHEDULE-AGENT=CLIENT (no scheduling)
* Added support for SCHEDULE-AGENT=NONE (no scheduling)

Please check changes locally because these java files are big and linting changed everything.
But:
* I have added SCHEDULE_AGENT on ZCalendar.ICalTok
* I have added ZCalendar.ScheduleAgent Enum
* Changed the Invite class to set scheduleAgent field on ZOrganizer and ZAttendee when receiving the Calendar request
* Modified OrganizerAutoScheduler to return email address of attendees that do not have SCHEDULE-AGENT=CLIENT or NONE (these should not be scheduled as message)
* Modified CalendarUser to support scheduleAgent field (ZOrganizer and ZAttendee are children of it)

### Tests
* See OrganizerAutoScheduleTest
* Tested with Outlook CDS: when SCHEDULE-AGENT=CLIENT Carbonio does not send invitation email, only Outlook.

### CODE REVIEW

File: common/src/main/java/com/zimbra/common/calendar/ZCalendar.java [hunk 1]
```diff
@@ -16,997 +21,1209 @@
...
+  public static final String sIcalVersion = "2.0";
+  public static final String sObsoleteVcalVersion = "1.0";
+
+  /**
+   * Values for SCHEDULE-AGENT param (https://www.rfc-editor.org/rfc/rfc6638.html#section-7.1)
+   *
+   * @author davidefrison
+   * @since 23.3.0
+   */
+  public static enum ScheduleAgent {
```
Comment (keshavbhatt) [inline common/src/main/java/com/zimbra/common/calendar/ZCalendar.java:41]: remove static keyword, its static by default.

File: store/src/main/java/com/zimbra/cs/dav/caldav/AutoScheduler.java [hunk 2]
```diff
@@ -42,573 +34,755 @@
...
+     * @param attendees meeting attendees
+     * @return list of email addresses
+     */
+    public List<Address> getRecipientsForAttendees(List<ZAttendee> attendees) {
+      if (attendees == null || attendees.isEmpty()) {
+        return Lists.newArrayListWithCapacity(0);
+      }
+      List<Address> recipients = Lists.newArrayListWithCapacity(attendees.size());
+      for (ZAttendee attendee : attendees) {
+        if (Objects.equals(attendee.getScheduleAgent(), "CLIENT")
```
Comment (keshavbhatt) [inline store/src/main/java/com/zimbra/cs/dav/caldav/AutoScheduler.java:622]: instead of hardcoding CLIENT and NONE (which quickly makes no sense) consider using already defined constants in ScheduleAgent enum.

Comment (frisonisland) [reply store/src/main/java/com/zimbra/cs/dav/caldav/AutoScheduler.java:622]: Thanks I added them for that very purpose, I forgot to use them :sweat_smile:

File: store/src/main/java/com/zimbra/cs/mailbox/calendar/CalendarUser.java [hunk 3]
```diff
@@ -5,312 +5,343 @@
...
+  private String mCn;
+  private String mSentBy;
+  private String mDir;
+  private String mLanguage;
+
+  public String getScheduleAgent() {
+    return scheduleAgent;
+  }
+
+  public CalendarUser setScheduleAgent(String scheduleAgent) {
```
Comment (keshavbhatt) [inline store/src/main/java/com/zimbra/cs/mailbox/calendar/CalendarUser.java:44]: consider using already defined ScheduleAgent enum as method parameter instead of string.

**Review by keshavbhatt — CHANGES_REQUESTED** [review]
I requested few changes to improve code quality.
Everything else looks good.

File: store/src/main/java/com/zimbra/cs/mailbox/calendar/CalendarUser.java (hunk 3 above)
Comment (frisonisland) [reply store/src/main/java/com/zimbra/cs/mailbox/calendar/CalendarUser.java:44]: Ok, the initial idea was just to use it to check values, as the schedule agent on CalendarUser is what we get from the request (if nothing fallbacks to "SERVER") and there could be corner cases like wrong case, etc. so by changing this we should take extra care of parsing the input value. I will change it.

Comment (frisonisland) [reply store/src/main/java/com/zimbra/cs/mailbox/calendar/CalendarUser.java:44]: I'm doing some changes, there will be more boilerplate basically

**Comment by frisonisland** [comment]
Reading https://www.rfc-editor.org/rfc/rfc5545 (https://www.rfc-editor.org/rfc/rfc6638.html extends it) property parameters value are intended to be case-insensitive but this is not covered yet, I will add changes.

File: store/src/main/java/com/zimbra/cs/mailbox/calendar/CalendarUser.java (hunk 3 above)
Comment (keshavbhatt) [reply store/src/main/java/com/zimbra/cs/mailbox/calendar/CalendarUser.java:44]: > Ok, the initial idea was just to use it to check values, as the schedule agent on CalendarUser is what we get from the request (if nothing fallbacks to "SERVER") and there could be corner cases like wrong case, etc. so by changing this we should take extra care of parsing the input value. For example the value "serVer" cannot be translated to ScheduleAgent, same for null string, do we still want to proceed?
> 
> The alternative is to just accept whatever comes from the client, if it does not match "CLIENT" or "NONE" then it sends the message.

I understand that using the string param will make it flexible but the idea is to follow the RFC. The RFC defines strict rules and conditions that must be followed. And to enforce them we must consider restricting to use constants defined in the ENUM. 

I think it will be simple to handle exceptional cases:
if SCHEDULE-AGENT param does not matches the strictly defined values, we can assume the parameter got wrong value and in that case user SERVER as suggested in RFC. I quote from (https://www.rfc-editor.org/rfc/rfc6638.html#section-7.1 [Page 42]):

>This property parameter MAY be specified on "ORGANIZER"
      or "ATTENDEE" iCalendar properties.  In the absence of this
      parameter, the value "SERVER" MUST be used for the default
      behavior.

Comment (frisonisland) [reply store/src/main/java/com/zimbra/cs/mailbox/calendar/CalendarUser.java:44]: Agree, also since it is case-insensitive it makes more sense at this point to map the value to a state of Schedule-Agent to avoid issues.

File: store/src/main/java/com/zimbra/cs/mailbox/calendar/CalendarUser.java [hunk 4]
```diff
@@ -37,16 +39,37 @@ public abstract class CalendarUser {
...
-    return this;
   }
 
-  private String scheduleAgent;
+  /**
+   * Converts a schedule agent value from string
+   *
+   * @param scheduleAgent
+   */
+  public void setScheduleAgentFromString(String scheduleAgent) {
```
Comment (keshavbhatt) [inline store/src/main/java/com/zimbra/cs/mailbox/calendar/CalendarUser.java:55]: Why this is required?

Comment (frisonisland) [reply store/src/main/java/com/zimbra/cs/mailbox/calendar/CalendarUser.java:55]: Because we need to translate the incoming request to ScheduleAgent for ZOrganzier and ZAttendee.
I could have done it in a private method of Invite.java class, but that would have meant to implement such method each time converting case-insensitive pops up, so I thought better add it to the enum.

Comment (frisonisland) [reply store/src/main/java/com/zimbra/cs/mailbox/calendar/CalendarUser.java:55]: Plus using .valueOf() parses only case-sensitive inputs and throws exception if provided value is not in enums so it does not cover the scenario.

Comment (keshavbhatt) [reply store/src/main/java/com/zimbra/cs/mailbox/calendar/CalendarUser.java:55]: What if this method gets arbitary string that is not covered by the enum? I dont see no default case. :confused:

Comment (frisonisland) [reply store/src/main/java/com/zimbra/cs/mailbox/calendar/CalendarUser.java:55]: It returns null at the end! I can add the default, np

Comment (keshavbhatt) [reply store/src/main/java/com/zimbra/cs/mailbox/calendar/CalendarUser.java:55]: This is throwing IllegalArgumentException. Please add a unit test for this method to confirm possible edge cases.

**Review by keshavbhatt — CHANGES_REQUESTED** [review]

**Review by keshavbhatt — APPROVED** [review]
lgtm.

**Review by Polpetta — APPROVED** [review]

**Comment by frisonisland** [comment]
I won't make other code changes since everyone approved, but will add some new tests to Invite class since it was never covered.
